### ISSUE TITLE
Graphql query filters with or / not operators, multiple filters

### ISSUE THREAD

**Issue by DevSide** [gatsbyjs/gatsby#22763]
## Summary

It looks like we cannot filter with **OR** or **NOT** operators in graphql queries, neither use multiple filters or combine query results properly. So I don't know exactly which feature it will be but at least being able to do **linear algebra**. Is there a **where** clause ?

### Basic example

Retrieve all entites where their fields match `(a && !b) || (!a && b)` in a single filter.
I guess the only way is to create two separate queries and combine them.

```graphql
allEntity (filter: { a: { eq: true }, b: { eq: false } }) {}
allEntity (filter: { a: { eq: false }, b: { eq: true } }) {}
```

### Motivation

Keeping UI rules query side only.
Avoid adding extra client code.

**Comment by freiksenet** [comment]
You can create a custom root resolver using [createResolvers](https://www.gatsbyjs.org/docs/schema-customization/#createresolvers-api). You can then combine the results inside the resolver and return combined results to the client.

**Comment by LekoArts** [comment]
Thank you for opening this!

We're marking this issue as answered and closing it for now but please feel free to comment here if you would like to continue this discussion. We also recommend heading over to our [communities](https://www.gatsbyjs.org/contributing/community/#where-to-get-support) if you have questions that are not bug reports or feature requests. We hope we managed to help and thank you for using Gatsby!

**Comment by DevSide** [comment]
closed ??

**Comment by tony** [comment]
@LekoArts @freiksenet Hi there!

Thank you for your courteous reply @freiksenet 

I'm wondering if this is a potential feature request or a documentation addition.`

I have been googling around and viewing the docs for examples of an equivalent to an `$or` query.

1. I don't believe I see anything on [createResolvers](https://www.gatsbyjs.org/docs/schema-customization/#createresolvers-api) that show an example of doing an equivalent to an "or" query or a combined result. If there is, please let me know!
2. I don't believe we're fully compatible with sift.  [Here is our page on what we support querying with sift in gatsby](https://www.gatsbyjs.org/docs/schema-sift/), and sift supports [`$or`](https://github.com/crcn/sift.js/#or) and [`$not`](https://github.com/crcn/sift.js/#not)

As for a combined query, the thing I have managed now is something like this:

```js
knownFor: {
  type: ["MyYaml"],
    async resolve(source: any, args: any, context: any, info: any) {
      // Todo find a more efficient way to do this, and only return a scalar value
      // firstOnly won't work
      const resultsQuery = context.nodeModel.runQuery({
        query: {
          filter: {
            author: {
              elemMatch: { id: { eq: source.id } },
            },
          },
        },
        type: "MyYaml",
      });
      const results = await resultsQuery;
      if (results) {
        return resultsQuery;
      }
      const resultsQuery2 = context.nodeModel.runQuery({
        query: {
          filter: {
            contributors: {
              elemMatch: { id: { eq: source.id } },
            },
          },
        },
        type: "MyYaml",
      });
      const results2 = await resultsQuery2;
      return resultsQuery2;
    },
  }
}
```

**Comment by DevSide** [comment]
It's interesting. I won't use `createResolvers` where the logic is too far from the components and their queries.

Here is some query apis which use OR / NOT operators:

graphcool
https://www.graph.cool/docs/reference/graphql-api/query-api-nia9nushae/#filtering-by-field

hasura
https://hasura.io/docs/1.0/graphql/manual/queries/query-filters.html#using-multiple-filters-in-the-same-query-and-or

drupal api
https://drupal-graphql.gitbook.io/graphql/queries/filters

contentstack api
https://www.contentstack.com/docs/developers/apis/graphql-content-delivery-api/#or-operator

**Comment by tony** [comment]
@DevSide
 
As a next step, should new issue(s) be made re-articulating the situation? Or this reopened?

In this case, is this a feature request (e.g. something not existing in gatsby yet)? If so, is it due to it causing potential overhead somehow? (or do we still need to find that out?)

**Comment by DevSide** [comment]
Well, I can reformulate this feature request with exemples. Unfortunately, I don't know if Gatsby team has already worked on a `where`/`conditions` filtering yet.

**Comment by tony** [comment]
@lekoarts: Would you like to re-open this?

**Comment by freiksenet** [comment]
Generally, we don't want to add any more functionality to our query language. It already exposes an API layer that's too complex. I feel that the combined query solution that @tony provided above is a good workaround.

**Comment by tony** [comment]
> Generally, we don't want to add any more functionality to our query language.

I don't want to push too hard, but isn't something like `OR` be a fair request?  Can you direct me to the part of the code where this happens? (Maybe it actually is complicated enough to be not possible, but I guess it's good to document it for the record, in case someone comes via Google / Searching the tracker)

>  I feel that the combined query solution that @tony provided above is a good workaround.

Would you have anything to offer on how to join the results of two queries inside a `resolve()` efficiently?  I'm not the best at wrangling promises. And also, can the be ran in parallel?

In my example, the runQuery is sort of acting as a fallback. I don't know if `runQuery`'s can actually be joined together.

Then, can we document the snippet example in the documentatios.

**Comment by scottwojan** [comment]
I have to agree this feels like baseline functionality for filtering capability. I just hit a wall due to not having OR capability.

**Comment by MANTENN** [comment]
I tried using unions then found out unions do not support arguments... 
```
{
  allContentfulType(filter: {campaign: {}}) {
    ... on Type {
      edges {
        node {
          id
          title
          campaign
        }
      }
    }
    ... on Type {
      edges {
        node {
          id
          title
          campaign
        }
      }
    }
  }
}
```

What you can do is use two queries with an alias and concatenate them on the results page:

```
{
  allContentfulType(filter: {campaign: { eq: null }}) {
    edges {
      node {
        id
        title
        campaign
      }
    }
  }
  defaultAllContentfulType: allContentfulType(filter: {campaign: { in: ["default"] }}) {
    edges {
      node {
        id
        title
        campaign
      }
    }
  }
}

```

**Comment by MANTENN** [comment]
I created a repo with a rough idea of how the plugin will work right here: https://github.com/MANTENN/gatsby-plugin-logic-filters/issues/1

**Comment by caloricpizza** [comment]
API
        .graphql(graphqlOperation(queries.listMessages, {
          channelID: "1",
          sortDirection: 'ASC',
          filter: { AND: [{ author: { eq: user }, receiver:{eq:receiver}},
          { receiver: { eq: user }, author:{eq:receiver}}]}
          
        }))
I coded like this.
![Screenshot_1](https://user-images.githubusercontent.com/74836776/104774492-017fca00-5734-11eb-81e9-d7b8fa72a649.png)
Error message like this. 
I believe you can help me.

**Comment by 3200pro** [comment]
I ran into this today. For now my quick fix was to write 2 matching queries that use a fragment to keep the redundancy down to a minimum.

Within my component I checked what type of products showed up after the filtering. In my case of need the OR operator this worked well:

```
productsByType: allSanitySiteProduct(filter: { productType: { elemMatch: { id: { eq: $id } } } }) {
      <fragment>
    }
    
productsByBrand: allSanitySiteProduct(filter: { productBrand: { elemMatch: { id: { eq: $id } } } }) {
      <fragment>
    }
```

`const products = data?.productsByType || data?.productsByBrand;`
